src/main/java/com/aws/greengrass/deployment/DeploymentService.java

```diff
@@ -437,35 +440,55 @@ private void createNewDeployment(Deployment deployment) {
         deploymentStatusKeeper.persistAndPublishDeploymentStatus(deployment.getId(), deployment.getDeploymentType(),
                                                                  JobStatus.IN_PROGRESS.toString(), new HashMap<>());
 
-        if (DEFAULT.equals(deployment.getDeploymentStage())
-                && DeploymentType.LOCAL.equals(deployment.getDeploymentType())) {
-            try {
-                copyRecipesAndArtifacts(deployment);
-            } catch (InvalidRequestException | IOException e) {
-                logger.atError().log("Error copying recipes and artifacts", e);
-                HashMap<String, String> statusDetails = new HashMap<>();
-                statusDetails.put("error", e.getMessage());
-                deploymentStatusKeeper.persistAndPublishDeploymentStatus(deployment.getId(),
-                        deployment.getDeploymentType(), JobStatus.FAILED.toString(), statusDetails);
-                return;
+        if (DEFAULT.equals(deployment.getDeploymentStage())) {
+
+            List<String> requiredCapabilities = deployment.getDeploymentDocumentObj().getRequiredCapabilities();
+            if (requiredCapabilities != null && !requiredCapabilities.isEmpty()) {
+                List<String> missingCapabilities = requiredCapabilities.stream()
+                        .filter(reqCapabilities -> !kernel.getSupportedCapabilities().contains(reqCapabilities))
+                        .collect(Collectors.toList());
+                if (!missingCapabilities.isEmpty()) {
+                    DeploymentResult result = new DeploymentResult(DeploymentStatus.FAILED_NO_STATE_CHANGE,
```

Comment:
don't you need to use persistAndPublishDeploymentStatus?

Reply:
persistAndPublishDeploymentStatus is called from finishCurrentDeployment